packages/web3-providers-ipc/src/index.ts

```diff
@@ -0,0 +1,193 @@
+import {
+	ConnectionNotOpenError,
+	DeferredPromise,
+	EthExecutionAPI,
+	InvalidClientError,
+	InvalidConnectionError,
+	JsonRpcId,
+	JsonRpcNotification,
+	JsonRpcResponse,
+	JsonRpcResponseWithError,
+	JsonRpcResponseWithResult,
+	JsonRpcResult,
+	Web3APIMethod,
+	Web3APIPayload,
+	Web3APIReturnType,
+	Web3APISpec,
+	Web3BaseProvider,
+	Web3BaseProviderCallback,
+	Web3BaseProviderStatus,
+} from 'web3-common';
+
+export default class IpcProvider<
+	API extends Web3APISpec = EthExecutionAPI,
+> extends Web3BaseProvider<API> {
+	private readonly _emitter: EventEmitter = new EventEmitter();
+
+	private readonly _socketPath: string;
+	private readonly _socket: Socket;
+	private _connectionStatus: Web3BaseProviderStatus;
+
+	private readonly _requestQueue: Map<JsonRpcId, DeferredPromise<unknown>>;
+
+	public constructor(socketPath: string) {
+		super();
+
+		this._connectionStatus = 'disconnected';
+		this._socketPath = socketPath;
+		this._socket = new Socket();
+
+		this._requestQueue = new Map<JsonRpcId, DeferredPromise<unknown>>();
+
+		this.connect();
+	}
+
+	public getStatus(): Web3BaseProviderStatus {
+		return this._connectionStatus;
+	}
+
+	/* eslint-disable class-methods-use-this */
+	public supportsSubscriptions(): boolean {
+		return true;
+	}
+
+	public on<T = JsonRpcResult>(
+		type: 'message' | 'connect' | 'disconnect' | string,
+		callback: Web3BaseProviderCallback<T>,
+	): void {
+		this._emitter.on(type, callback);
+	}
+
+	public once<T = JsonRpcResult>(type: string, callback: Web3BaseProviderCallback<T>): void {
+		this._emitter.once(type, callback);
+	}
+
+	public removeListener(type: string, callback: Web3BaseProviderCallback): void {
+		this._emitter.removeListener(type, callback);
+	}
+
+	public connect(): void {
+		if (!existsSync(this._socketPath)) {
+			throw new InvalidClientError(this._socketPath);
+		}
+
+		try {
+			this._addSocketListeners();
+			this._socket.connect({ path: this._socketPath });
+		} catch (e) {
+			throw new InvalidConnectionError(this._socketPath);
+		}
+	}
+
+	public disconnect(): void {
+		this._requestQueue.clear();
+		this._removeSocketListeners();
+		this._socket.end();
+	}
+
+	public reset(): void {
+		this._requestQueue.clear();
+
+		this._removeSocketListeners();
+		this._addSocketListeners();
+	}
+
+	// eslint-disable-next-line @typescript-eslint/require-await
+	public async request<
+		Method extends Web3APIMethod<API>,
+		ResponseType = Web3APIReturnType<API, Method>,
+	>(request: Web3APIPayload<API, Method>): Promise<JsonRpcResponse<ResponseType>> {
+		if (this._socket === undefined) throw new Error('IPC connection is undefined');
+
+		if (request.id === undefined) throw new Error('Request Id not defined');
+
+		if (this.getStatus() !== 'connected') {
+			throw new ConnectionNotOpenError();
+		}
+
+		try {
+			this._socket.write(JSON.stringify(request));
+			const defPromise = new DeferredPromise<JsonRpcResponse<ResponseType>>();
+			this._requestQueue.set(request.id, defPromise);
```

Comment:
```suggestion
			const defPromise = new DeferredPromise<JsonRpcResponse<ResponseType>>();
			this._requestQueue.set(request.id, defPromise);
			this._socket.write(JSON.stringify(request));
```

Should it be written this way to avoid possible race conditions?